be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/RollCallHandler.scala

```diff
@@ -72,8 +72,59 @@ sealed class RollCallHandler(dbRef: => AskableActorRef) extends MessageHandler {
   }
 
   def handleCloseRollCall(rpcMessage: JsonRpcRequest): GraphMessage = {
+    // We first create the three (independent) futures
     val ask: Future[GraphMessage] = dbAskWritePropagate(rpcMessage)
-    Await.result(ask, duration) match {
+
+    // Creates a channel for each attendee (of name /root/lao_id/social/PublicKeyAttendee), returns a Future[GraphMessage]
+    def createAttendeeChannels(data: CloseRollCall, rpcMessage: JsonRpcRequest): Future[GraphMessage] = {
+      val listAttendeeChannels: List[(Channel, ObjectType.ObjectType)] = data.attendees.map(attendee => (generateSocialChannel(rpcMessage.getParamsChannel, attendee), ObjectType.CHIRP))
+      val askCreateChannels: Future[GraphMessage] = (dbActor ? DbActor.CreateChannelsFromList(listAttendeeChannels)).map{
+        case DbActorAck() => Left(rpcMessage)
+        case DbActorNAck(code, description) => Right(PipelineError(code, description, rpcMessage.id))
+        case _ => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, unknownAnswer, rpcMessage.id))
+      }
+      askCreateChannels
+    }
+
+    val askOldData = dbActor ? DbActor.ReadLaoData(rpcMessage.getParamsChannel)
+
+    rpcMessage.getParamsMessage match {
+      case Some(message: Message) =>
+        val data: CloseRollCall = message.decodedData.get.asInstanceOf[CloseRollCall]
+        val askCreateChannels: Future[GraphMessage] = createAttendeeChannels(data, rpcMessage)
+
+        // We create a new Future (tuple of Futures) to remove the obligatory sequential execution
+        // TODO: A possible addition would be to handle the error messages better
+        val resFuture: Future[GraphMessage] = (for{
+          f1 <- ask
+          f2 <- askOldData
+          f3 <- askCreateChannels
+        } yield(f1, f2, f3)).map{
+          case (Left(_), DbActorReadLaoDataAck(Some(_)), Left(_)) => Left(rpcMessage)
+          case (Left(_), DbActorNAck(code, description), Left(_)) => Right(PipelineError(code, description, rpcMessage.id))
```

Comment:
In this case, wouldn't we have created the channels already, even though the LAO doesn't exist? 
My understanding is that the `flatMap` / comprehension would go through each step, as `dbActorNAck` is not a failure.

Reply:
Pierluca is correct in this case. I want to also bring another point... In this case, assuming that the future called `ask` fails (i.e. a true exception, not a Success(DbActorNack)), then **the 2nd features (i.e. askOldData) will still be executed**. Let me show you one last example

Assuming our database is `dbState` (you can imagine adding something to `dbState` as "writing something to be2's database"), then watch what happens when I define ff1 (that fails) and ff2 (that succeeds). No matter what ff1 does, ff2 will still accomplish his task. 

<img width="801" alt="Screenshot 2022-02-11 at 05 06 03" src="https://user-images.githubusercontent.com/44499915/153536434-33eb1809-e410-43da-a079-9efd8173f49e.png">



In the project, if ff1 creates an LAO (and fails) and ff2 updates its information, then ff2 will update information on a LAO that does not exist

Reply:
Oh right I was a bit confused on what the issue was and I get it now. However, updating a non-existing LAO should yield an error and therefore not break the code, or am I wrong and something should be changed?